### PR TITLE
Fix dirlist method return type

### LINKED ISSUE #93: `WP_Filesystem_Base::dirlist` returns an array<array>
### Types from PHPDocs
```php
@return array|false {                                                                                  
    Array of files. False if unable to list directory contents.                                        
                                                                                                       
    @type string $name        Name of the file or directory.                                           
    @type string $perms       *nix representation of permissions.                                      
    @type string $permsn      Octal representation of permissions.                                     
    @type string $owner       Owner name or ID.                                                        
    @type int    $size        Size of file in bytes.                                                   
    @type int    $lastmodunix Last modified unix timestamp.                                            
    @type mixed  $lastmod     Last modified month (3 letter) and day (without leading 0).              
    @type int    $time        Last modified time.                                                      
    @type string $type        Type of resource. 'f' for file, 'd' for directory.                       
    @type mixed  $files       If a directory and `$recursive` is true, contains another array of files.
}       
```
### Types inferred in stubs
```php
 @phpstan-return false|array{          
   name: string,                       
   perms: string,                      
   permsn: string,                     
   owner: string,                      
   size: int,                          
   lastmodunix: int,                   
   lastmod: mixed,                     
   time: int,                          
   type: string,                       
   files: mixed,                       
 }                                     
```       
### Actual return type
```
false|array<string, array<{          
   name: string,                       
   perms: string,                      
   permsn: string,                     
   owner: string,                      
   size: int,                          
   lastmodunix: int,                   
   lastmod: mixed,                     
   time: int,                          
   type: string,                       
   files: mixed,                       
 }>
```

Notice the code returns a multidimensional array, not a single associative array.

https://github.com/WordPress/WordPress/blob/e3d1a1291fd3bbfdca3af409e8007d35f2109fa4/wp-admin/includes/class-wp-filesystem-direct.php#L670

Same issue in classes which use the base interface:
1. WP_Filesystem_Direct
2. WP_Filesystem_FTPext
3. WP_Filesystem_MockFS
4. WP_Filesystem_SSH2
5. WP_Filesystem_ftpsockets

### PR DESCRIPTION
**Pull request by szepeviktor**
@lipemat I could not resist. I am sorry.
Could you please review this PR?

Fixes #93

### CODE REVIEW

**Review by johnbillion — CHANGES_REQUESTED** [review]
This still isn't quite right as the core docs documented some of the array elements incorrectly.

PR for core here: https://github.com/WordPress/wordpress-develop/pull/4407

**Comment by szepeviktor** [comment]
Okay. Copying your types ...

File: functionMap.php [hunk 1]
```diff
@@ -123,4 +124,9 @@
     'wp_register' => ['($display is true ? void : string)'],
     'wp_title' => ['($display is true ? void : string)'],
     'wp_script_is' => [null, 'status' => "'enqueued'|'registered'|'queue'|'to_do'|'done'"],
+    'WP_Filesystem_Direct::dirlist' => [$filesystemDirlistReturnType],
+    'WP_Filesystem_FTPext::dirlist' => [$filesystemDirlistReturnType],
+    'WP_Filesystem_MockFS::dirlist' => [$filesystemDirlistReturnType],
```
Comment (johnbillion) [inline functionMap.php:129]:
```suggestion
    'WP_Filesystem_Base::dirlist' => [$filesystemDirlistReturnType],
```

**Review by johnbillion — CHANGES_REQUESTED** [review]

File: functionMap.php (hunk 1 above)
Comment (szepeviktor) [reply functionMap.php]: Thank you.

File: functionMap.php [hunk 2]
```diff
@@ -8,6 +8,7 @@
 $cronArgsType = 'list<mixed>';
 $wpWidgetRssFormArgsType = 'array{number: int, error: bool, title?: string, url?: string, items?: int, show_summary?: int, show_author?: int, show_date?: int}';
 $wpWidgetRssFormInputType = 'array{title?: bool, url?: bool, items?: bool, show_summary?: bool, show_author?: bool, show_date?: bool}';
+$filesystemDirlistReturnType = 'false|array<string, array{name: string, perms: string, permsn: string, owner: string|false, size: int|string|false, lastmodunix: int|string|false, lastmod: string|false, time: string|false, type: string, files: array}>';
```
Comment (johnbillion) [inline functionMap.php:11]:
```suggestion
$filesystemDirlistReturnType = 'false|array<string, array{name: string, perms: string, permsn: string, owner: string|false, size: int|string|false, lastmodunix: int|string|false, lastmod: string|false, time: string|false, type: 'f'|'d'|'l', files: array}>';
```

**Review by johnbillion — CHANGES_REQUESTED** [review]

File: functionMap.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (szepeviktor) [reply functionMap.php:11]: Thank you.
Where do group and number come from?

Comment (szepeviktor) [reply functionMap.php:11]: Goes into mater in a minute.

[the remark this answers is not in the document]
Comment (szepeviktor) [reply functionMap.php:11]: I see!!

[the remark this answers is not in the document]
Comment (szepeviktor) [reply functionMap.php:11]: Fixes in https://github.com/php-stubs/wordpress-stubs/commit/ed9ad16fa0ed1493cb39668242515041e1a326c8
Thank you.

**Pull request merged**